packages/langium/src/utils/collections.ts

```diff
@@ -0,0 +1,139 @@
+/******************************************************************************
+ * Copyright 2021 TypeFox GmbH
+ * This program and the accompanying materials are made available under the
+ * terms of the MIT License, which is available in the project root.
+ ******************************************************************************/
+
+import { Reduction, Stream, stream } from './stream';
+
+/**
+ * A multimap is a variation of a Map that has potentially multiple values for every key.
+ */
+export class MultiMap<K, V> {
+
+    private map = new Map<K, V[]>();
+
+    /**
+     * The total number of values in the multimap.
+     */
+    get size(): number {
+        return Reduction.sum(stream(this.map.values()).map(a => a.length));
+    }
+
+    /**
+     * Clear all entries in the multimap.
+     */
+    clear(): void {
+        this.map.clear();
+    }
+
+    /**
+     * Operates differently depending on whether a `value` is given:
+     *  * With a value, this method deletes the specific key / value pair from the multimap.
+     *  * Without a value, all values associated with the given key are deleted.
+     *
+     * @returns `true` if a value existed and has been removed, or `false` if the specified
+     *     key / value does not exist.
+     */
+    delete(key: K, value?: V): boolean {
+        if (value === undefined) {
+            return this.map.delete(key);
+        } else {
+            const values = this.map.get(key);
+            if (values) {
+                const index = values.indexOf(value);
+                if (index >= 0) {
+                    if (values.length === 1) {
+                        this.map.delete(key);
+                    } else {
+                        values.splice(index, 1);
+                    }
+                    return true;
+                }
+            }
+            return false;
+        }
+    }
+
+    /**
+     * Returns an array of all values associated with the given key. If no value exists,
+     * an empty array is returned.
+     *
+     * _Note:_ The returned array is assumed not to be modified. Use the `set` method to add a
+     * value and `delete` to remove a value from the multimap.
+     */
+    get(key: K): readonly V[] {
+        return this.map.get(key) ?? [];
```

Comment:
I had to look twice until I spotted the readonly :)

Reply:
I used `ReadonlyArray<V>` first, but eslint complained about that :-)